### PR TITLE
[BOLT] Allow pass-through blocks in YAMLProfileReader

### PR DESCRIPTION
**Pull request by aaupov**
Align YAMLProfileReader with DataReader in handling non-matching edges.

Allow pass-through blocks in non-matching edges:
match the profile edge A -> C to the CFG with blocks A -> B -> C.

This case arises when profiling BOLTed binaries with block B
removed.

Test Plan: Added profile-passthrough-block.test

### CODE REVIEW

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 1]
```diff
@@ -218,17 +218,29 @@ bool YAMLProfileReader::parseFunctionProfile(
...
 
-      BinaryBasicBlock &SuccessorBB = *Order[YamlSI.Index];
-      if (!BB.getSuccessor(SuccessorBB.getLabel())) {
-        if (opts::Verbosity >= 1)
-          errs() << "BOLT-WARNING: no successor for block " << BB.getName()
-                 << " that matches index " << YamlSI.Index << " or block "
-                 << SuccessorBB.getName() << '\n';
-        ++MismatchedEdges;
-        continue;
+      BinaryBasicBlock *ToBB = Order[YamlSI.Index];
```
Comment (ayermolo) [inline bolt/lib/Profile/YAMLProfileReader.cpp:221]: Cast directly to reference?

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 2]
```diff
@@ -218,17 +218,29 @@ bool YAMLProfileReader::parseFunctionProfile(
...
+        if (FTSuccessor && FTSuccessor->succ_size() == 1 &&
+            FTSuccessor->getSuccessor(ToBB->getLabel())) {
+          BinaryBasicBlock::BinaryBranchInfo &FTBI =
+              FTSuccessor->getBranchInfo(*ToBB);
+          FTBI.Count += YamlSI.Count;
+          FTBI.MispredictedCount += YamlSI.Mispreds;
+          ToBB = FTSuccessor;
+        } else {
+          if (opts::Verbosity >= 1)
+            errs() << "BOLT-WARNING: no successor for block " << BB.getName()
```
[the remark this answers is not in the document]
Comment (aaupov) [reply bolt/lib/Profile/YAMLProfileReader.cpp:235]: Not in the Profile component (yet). We don't have error propagating interfaces there yet.

File: bolt/lib/Profile/YAMLProfileReader.cpp (hunk 1 above)
Comment (aaupov) [reply bolt/lib/Profile/YAMLProfileReader.cpp:221]: We reassign ToBB down below

File: bolt/test/X86/profile-passthrough-block.test [hunk 3]
```diff
@@ -0,0 +1,67 @@
+## This reproduces a bug with BOLT setting incorrect discriminator for
+## secondary entry points in YAML profile.
```
Comment (maksfb) [inline bolt/test/X86/profile-passthrough-block.test:2]: Change the comment.

File: bolt/lib/Profile/YAMLProfileReader.cpp [hunk 4]
```diff
@@ -218,17 +218,29 @@ bool YAMLProfileReader::parseFunctionProfile(
...
-        if (opts::Verbosity >= 1)
-          errs() << "BOLT-WARNING: no successor for block " << BB.getName()
-                 << " that matches index " << YamlSI.Index << " or block "
-                 << SuccessorBB.getName() << '\n';
-        ++MismatchedEdges;
-        continue;
+      BinaryBasicBlock *ToBB = Order[YamlSI.Index];
+      if (!BB.getSuccessor(ToBB->getLabel())) {
+        // Allow for BOLT-removed passthrough blocks to align with DataReader
+        // behavior.
```
Comment (maksfb) [inline bolt/lib/Profile/YAMLProfileReader.cpp:224]: The change in `DataReader` was introduced for backwards compatibility with old profiles (and we can remove it). This PR serves a different purpose.

**Review by maksfb — COMMENTED** [review]
Is it possible that we remove more blocks in the chain? E.g. A -> [ B -> C -> ] D?

**Comment by aaupov** [comment]
> Is it possible that we remove more blocks in the chain? E.g. A -> [ B -> C -> ] D?

Yes. Neither DataReader nor YAMLProfileReader will recognize that.

**Comment by maksfb** [comment]
> > Is it possible that we remove more blocks in the chain? E.g. A -> [ B -> C -> ] D?
> 
> Yes. Neither DataReader nor YAMLProfileReader will recognize that.

It shouldn't matter for the `DataReader` with new profiles.

File: bolt/lib/Profile/YAMLProfileReader.cpp (hunk 4 above)
Comment (aaupov) [reply bolt/lib/Profile/YAMLProfileReader.cpp]: Rephrased the comment.

**Review by maksfb — APPROVED** [review]

**Pull request merged**